server/swap/swap_test.go

```diff
@@ -1610,3 +1610,123 @@ func TestCancel(t *testing.T) {
 	}
 }
+
+func TestTxMonitored(t *testing.T) {
+	sendBlock := func(node *TAsset) {
+		node.bChan <- 1
+		tickMempool()
+	}
+
+	makerSell := true
+	set := tPerfectLimitLimit(uint64(1e8), uint64(1e8), makerSell)
+	matchInfo := set.matchInfos[0]
+	rig := tNewTestRig(matchInfo)
+	rig.swapper.Negotiate([]*order.MatchSet{set.matchSet})
+	ensureNilErr := makeEnsureNilErr(t)
+	//mustBeError := makeMustBeError(t)
+	maker, taker := matchInfo.maker, matchInfo.taker
+
+	var makerLockedAsset, takerLockedAsset uint32
+	if makerSell {
+		makerLockedAsset = matchInfo.match.Maker.Base()  // maker sell locks base asset
+		takerLockedAsset = matchInfo.match.Taker.Quote() // taker buy locks quote asset
+	} else {
+		makerLockedAsset = matchInfo.match.Maker.Quote() // maker sell locks base asset
+		takerLockedAsset = matchInfo.match.Taker.Base()  // taker buy locks quote asset
+	}
+
+	tracker := rig.getTracker()
+	if tracker.Status != order.NewlyMatched {
+		t.Fatalf("match not marked as NewlyMatched: %d", tracker.Status)
+	}
+
+	// Maker acks match and sends swap tx.
+	ensureNilErr(rig.ackMatch_maker(true))
+	ensureNilErr(rig.sendSwap_maker(true))
+	makerContractTx := rig.matchInfo.db.makerSwap.coin.TxID()
+	if !rig.swapper.TxMonitored(maker.acct, makerLockedAsset, makerContractTx) {
+		t.Errorf("maker contract %s (asset %d) was not monitored",
+			makerContractTx, makerLockedAsset)
+	}
+
+	if tracker.Status != order.MakerSwapCast {
+		t.Fatalf("match not marked as MakerSwapCast: %d", tracker.Status)
+	}
+	sendBlock(rig.abcNode)
+	sendBlock(rig.xyzNode)
+
+	// For the taker, there must be two acknowledgements before broadcasting the
+	// swap transaction, the match ack and the audit ack.
+	ensureNilErr(rig.ackMatch_taker(true))
+	ensureNilErr(rig.auditSwap_taker())
+	ensureNilErr(rig.ackAudit_taker(true))
+	ensureNilErr(rig.sendSwap_taker(true))
+
+	takerContractTx := rig.matchInfo.db.takerSwap.coin.TxID()
+	if !rig.swapper.TxMonitored(taker.acct, takerLockedAsset, takerContractTx) {
+		t.Errorf("taker contract %s (asset %d) was not monitored",
+			takerContractTx, takerLockedAsset)
+	}
+
+	if tracker.Status != order.TakerSwapCast {
+		t.Fatalf("match not marked as TakerSwapCast: %d", tracker.Status)
+	}
+	sendBlock(rig.abcNode)
+	sendBlock(rig.xyzNode)
+
+	ensureNilErr(rig.auditSwap_maker())
+	ensureNilErr(rig.ackAudit_maker(true))
+
+	// Set the number of confirmations on the contracts.
+	//tracker.makerStatus.swapTime
+	// tracker.makerStatus.swapConfirmed = time.Now()
+	// tracker.takerStatus.swapConfirmed = time.Now()
+	matchInfo.db.makerSwap.coin.setConfs(int64(rig.abc.SwapConf))
+	matchInfo.db.takerSwap.coin.setConfs(int64(rig.xyz.SwapConf))
+
+	// send a block through for either chain to trigger a swap check.
+	sendBlock(rig.abcNode)
+	sendBlock(rig.xyzNode)
+
+	if rig.swapper.TxMonitored(maker.acct, makerLockedAsset, makerContractTx) {
+		t.Errorf("maker contract %s (asset %d) was still monitored",
```

Comment:
Ready for review, but this test is failing.  My expectation is that it the maker contract should not be monitored at this point.  I'm having a hard time determining the cause, whether it is timing of the test harness or another issue, but the swap time seems to still be Zero.

Reply:
The swap is never marked as confirmed at https://github.com/chappjc/dcrdex/blob/c3fe624b2a534bd9adb2abaff8394e60d8d309ee/server/swap/swap.go#L370

because when you sent blocks through while the status was `MakerSwapCast`, you had not yet used `setConfs`. 

I hesitate to call this a bug. If the taker broadcasts their swap transaction before the the server is aware of the confirmations on the maker's swap, the server should accept the transaction and move forward. The alternative is for the server to reject the taker's `init` messages until the server knows the maker's swap has the requisite confirmations, which I like less.

Before `TxMonitored`, the maker's `(swapStatus).swapConfirmed` field was ignored once the status became `MakerSwapCast`, and the timestamp is not stored in the database, so there was no reason to continue looking for it. I guess now there is though, so `processBlock` might need some additional logic.